addon/serializers/json.js

```diff
@@ -843,14 +843,35 @@ var JSONSerializer = Serializer.extend({
   /**
     Check if the given hasMany relationship should be serialized
 
-    @method _shouldSerializeHasMany
+    @method shouldSerializeHasMany
+    @param {DS.Snapshot} snapshot
+    @param {String} key
+    @param {String} relationshipType
+    @return {boolean} true if the hasMany relationship should be serialized
+  */
+  shouldSerializeHasMany(snapshot, key, relationship) {
+    return this.__shouldSerializeHasMany(snapshot, key, relationship);
+  },
+
+  /**
+    Check if the given hasMany relationship should be serialized
+
+    @method __shouldSerializeHasMany
     @private
     @param {DS.Snapshot} snapshot
     @param {String} key
     @param {String} relationshipType
     @return {boolean} true if the hasMany relationship should be serialized
   */
-  _shouldSerializeHasMany(snapshot, key, relationship) {
+  __shouldSerializeHasMany(snapshot, key, relationship) {
+    if (this._shouldSerializeHasMany) {
+      deprecate('The private method _shouldSerializeHasMany has been promoted to the public API. Please enable feature flag `ds-check-should-serialize-relationships` and remove the underscore to use the public shouldSerializeHasMany method.', false, {
+        id: 'ds.serializer.private-should-serialize-has-many',
+        until: '3.0.0'
+      });
```

Comment:
The `deprecate` also needs to be behind a `if (!isEnabled("ds-check-should-serialize-relationships")) {` so we don't log the deprecation if the feature is not enabled..